packages/core/bootstrap/src/lib/ws/epics.ts

```diff
@@ -0,0 +1,183 @@
+import {
+  endWith,
+  filter,
+  map,
+  mergeMap,
+  startWith,
+  takeUntil,
+  tap,
+  withLatestFrom,
+} from 'rxjs/operators'
+import { webSocket } from 'rxjs/webSocket'
+import WebSocket from 'ws'
+import {
+  connect,
+  connected,
+  disconnect,
+  disconnected,
+  subscribe,
+  subscribed,
+  unsubscribe,
+  unsubscribed,
+  messageReceived,
+  WSConfigPayload,
+  WSSubscriptionPayload,
+} from './actions'
+import {
+  ws_connection_active,
+  ws_message_total,
+  ws_subscription_active,
+  ws_subscription_total,
+} from './metrics'
+
+export const connectEpic: Epic<AnyAction, AnyAction, any, any> = (action$, state$) =>
+  action$.pipe(
+    filter(connect.match),
+    map(({ payload }) => ({ payload, connectionKey: payload.config.connectionInfo.key })),
+    // check if the connection already exists, then noop
+    withLatestFrom(state$),
+    filter(([{ connectionKey }, state]) => {
+      // if connection does not exist, then continue
+      return !state.ws.connections.active[connectionKey]
+    }),
+    // on a connect action being dispatched, open a new WS connection if one doesn't exist yet
+    mergeMap(([{ connectionKey, payload }]) => {
+      const { url, protocol } = payload.config.connectionInfo
+      const openObserver = new Subject()
+      const closeObserver = new Subject()
+      const WebSocketCtor = WebSocket
+      const wsSubject = webSocket({
+        url,
+        protocol,
+        openObserver,
+        closeObserver,
+        WebSocketCtor: WebSocketCtor as any, // TODO: fix types don't match
+      })
+
+      const { config } = payload
+      // Stream of WS connected & disconnected events
+      const open$ = openObserver.pipe(map((event) => connected({ event, config } as any)))
+      const close$ = closeObserver.pipe(map((event) => disconnected({ event, config } as any)))
+
+      // Close the WS connection on disconnect
+      const disconnect$ = action$.pipe(
+        filter(disconnect.match),
+        filter(({ payload }) => payload.connectionInfo.key === connectionKey),
+        tap(() => wsSubject.closed || wsSubject.complete()),
+        filter(() => false), // do not duplicate events
+      )
+
+      // Multiplex subscriptions
+      const subscription$ = action$.pipe(
+        filter(subscribe.match),
+        filter(({ payload }) => payload.connectionInfo.key === connectionKey),
+        map(({ payload }) => ({ payload, subscriptionKey: payload.subscriptionInfo.key })),
+        withLatestFrom(state$),
+        filter(([{ subscriptionKey }, state]) => {
+          // if subscription does not exist, then continues
+          return !state.ws.subscriptions.active[subscriptionKey]
+        }),
+        // on a subscribe action being dispatched, open a new WS subscription if one doesn't exist yet
+        mergeMap(([{ subscriptionKey, payload }]) => {
+          // Helper function to build actions
+          const _wsSubscriptionPayload = (message: any) => ({
```

Comment:
Underscores mean unused in JS-land, we should use standing camelCase here :), if this function ever actually becomes unused and we forget to remove it, both tsc and eslint wont warn us because of the underscore convention